### ISSUE TITLE
Open issue: relax requirement that type be enumerable to participate in collection expressions

### ISSUE THREAD

**Issue by jcouv** [dotnet/csharplang#7744]
It is presently not possible to declare a collection type that doesn't implement some `IEnumerable` interface.
We should consider allowing types marked with `[CollectionBuilder(...)]` to not be enumerable (and thus not have an iterator type).

I ran into this issue trying to make a roslyn type (`CSharpTestSource`) usable with collection expressions. This required a dummy implementation of an `IEnumerable` interface despite that interface playing little role for compiling this code.

To achieve this, we would remove the requirement from the `[CollectionBuilder]` [section](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#create-methods):
> The collection type must have an iteration type.

And we would instead use the type from the parameter type of the `Create` method, ie. `E` in `ReadOnlySpan<E>`. All elements of the collection need to an implicit conversion to that type.

Relates to https://github.com/dotnet/csharplang/issues/5354
FYI @cston @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
Not sure how I feel about this.  

Why not make it enumerable?


Seems very weird that you found use a collection to create it, but then not be able to iterate it.

**Comment by CyrusNajmabadi** [comment]
I guess the same feelings existed for collection expressions :-). I'll look at the example though!

**Comment by CyrusNajmabadi** [comment]
> We explicitly (no pun intended) excluded that when we did not make a collection expression conversion a standard implicit conversion.

Yup yup. But I'm curious if we might reconsider given some real world use case. 

That said, CSharpSources seems, to me, to indicate a collection type. So not being enumerable is frankly kinda strange. I'll have to look and see how you guys actually end up using this.

**Comment by 333fred** [comment]
> I'll have to look and see how you guys actually end up using this.

We basically always realize it to a flattened array when we actually go to access it.

**Comment by CyrusNajmabadi** [comment]
so you have a collection type which is like an array, but it's all wonky on conversions to/from an array.  go tit.  :)

not sure how i feel about this case.  seems very different than basically like everything else out there.  note that i view collection-exprs as a common way to create collections.  Not a nice syntax to create *any* arbitrary type.  But i am open to this sort of relaxation because... well... it doesn't seem like it would harm anything.

**Comment by RikkiGibson** [comment]
I thought one intended use of CollectionBuilder was to allow creation of ref structs from collection-exprs. Is that not the case? [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAOgCUBXGAFyjAFMKBhAezAAcoAbZgCcAykIBuUAMbMAzgG4CBANqc+AyYw4wAQjX4IhACnoBPbsw4AzQ1WEBKOCRgQWVm8PaDmEeszt2AXQJ0AGYSL0sSGXpBGg0SWxIAbwISNLIwxLYvH2YbbwQAeRg+E2FuCBgAHlh6AD4SKF8wGTsSAF4G+gALQQ4AdycaNQBCBXx0jLJsADYyABYSAFlDNpSJyfTEwRlsDpIlbEdMRxCA8cmAXwJLoA==)

```cs
using System;
using System.Runtime.CompilerServices;

[CollectionBuilder(typeof(RS), nameof(RS.Create))]
public ref struct RS {
    public RS Create(ReadOnlySpan<int> items) => throw null!;
    public static void M() {
        RS rs1 = [1, 2, 3]; // error CS9188: 'RS' has a CollectionBuilderAttribute but no element type.
    }
}
```

**Comment by timcassell** [comment]
It would also be nice to relax the restriction for old style collection initializer syntax. An example where I wrote a class for a write-only collection initializer (before collection expressions existed): https://github.com/timcassell/ProtoBenchmarkHelpers#BenchmarkThreadHelper

**Comment by RikkiGibson** [comment]
Old collection initializers had a requirement to implement `IEnumerable` because without it, it was felt to be unclear if just an `Add(T)` taking a parameter of some type is intended for adding an element to a collection, or for e.g. performing arithmetic.

I think a type with `[CollectionBuilder]` not needing to be enumerable is a bit easier of a sell, as adding the attribute to the type very clearly indicates the associated Create method is intended for creating a collection. And folks who are interested in creating "write-only" collections should be able to onboard to it and get the experience they want.

**Comment by gulbanana** [comment]
> There was a community request for this exact issue description at [#7398 (reply in thread)](https://github.com/dotnet/csharplang/discussions/7398#discussioncomment-7010257).

Yep, I'd like to use CollectionBuilder (or some equivalent api) on what is basically a _writeonly_ collection. This type is an ORM query, which default-constructed and have members added, but you can't *iterate* them until you've performed an async load or save.
